crates/wasmtime/src/component/func.rs

```diff
@@ -241,4 +249,342 @@ impl Func {
 
         Ok(())
     }
+
+    /// Get the parameter types for this function.
+    pub fn params(&self, store: impl AsContext) -> Box<[Type]> {
+        let data = &store.as_context()[self.0];
+        data.types[data.ty]
+            .params
+            .iter()
+            .map(|(_, ty)| Type::from(ty, &data.types))
+            .collect()
+    }
+
+    /// Invokes this function with the `params` given and returns the result.
+    ///
+    /// The `params` here must match the type signature of this `Func`, or this will return an error. If a trap
+    /// occurs while executing this function, then an error will also be returned.
+    // TODO: say more -- most of the docs for `TypedFunc::call` apply here, too
+    pub fn call(&self, mut store: impl AsContextMut, args: &[Val]) -> Result<Val> {
+        let store = &mut store.as_context_mut();
+
+        let params;
+        let result;
+
+        {
+            let data = &store[self.0];
+            let ty = &data.types[data.ty];
+
+            if ty.params.len() != args.len() {
+                bail!(
+                    "expected {} argument(s), got {}",
+                    ty.params.len(),
+                    args.len()
+                );
+            }
+
+            params = ty
+                .params
+                .iter()
+                .zip(args)
+                .map(|((_, ty), arg)| {
+                    let ty = Type::from(ty, &data.types);
+
+                    ty.check(arg).context("type mismatch with parameters")?;
+
+                    Ok(ty)
+                })
+                .collect::<Result<Vec<_>>>()?;
+
+            result = Type::from(&ty.result, &data.types);
+        }
+
+        let param_count = params.iter().map(|ty| ty.flatten_count()).sum::<usize>();
+        let result_count = result.flatten_count();
+
+        self.call_raw(
+            store,
+            args,
+            |store, options, args, dst: &mut MaybeUninit<[ValRaw; MAX_STACK_PARAMS]>| {
+                if param_count > MAX_STACK_PARAMS {
+                    self.store_args(store, &options, &params, args, dst)
+                } else {
+                    args.iter()
+                        .try_for_each(|arg| arg.lower(store, &options, dst, 0))
```

Comment:
This seems like it may not be right for multiple arguments since the offset may always be zero?